flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializerSnapshot.java

```diff
@@ -124,11 +124,40 @@ void readSnapshot(int readVersion, DataInputView in, ClassLoader userCodeClassLo
      * program's serializer re-serializes the data, thus converting the format during the restore
      * operation.
      *
+     * @deprecated This method has been replaced by {@link TypeSerializerSnapshot
+     *     #resolveSchemaCompatibility(TypeSerializerSnapshot)}.
      * @param newSerializer the new serializer to check.
      * @return the serializer compatibility result.
      */
-    TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(
-            TypeSerializer<T> newSerializer);
+    @Deprecated
+    default TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(
+            TypeSerializer<T> newSerializer) {
+        return newSerializer.snapshotConfiguration().resolveSchemaCompatibility(this);
+    }
+
+    /**
+     * Checks current serializer's compatibility to read data written by the prior serializer.
+     *
+     * <p>When a checkpoint/savepoint is restored, this method checks whether the serialization
+     * format of the data in the checkpoint/savepoint is compatible for the format of the serializer
+     * used by the program that restores the checkpoint/savepoint. The outcome can be that the
+     * serialization format is compatible, that the program's serializer needs to reconfigure itself
+     * (meaning to incorporate some information from the TypeSerializerSnapshot to be compatible),
+     * that the format is outright incompatible, or that a migration needed. In the latter case, the
+     * TypeSerializerSnapshot produces a serializer to deserialize the data, and the restoring
+     * program's serializer re-serializes the data, thus converting the format during the restore
+     * operation.
+     *
+     * <p>This method must be implemented to clarify the compatibility. See FLIP-263 for more
+     * details.
+     *
+     * @param oldSerializerSnapshot the old serializer snapshot to check.
+     * @return the serializer compatibility result.
+     */
+    default TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(
+            TypeSerializerSnapshot<T> oldSerializerSnapshot) {
+        return TypeSerializerSchemaCompatibility.incompatible();
```

Comment:
For users' existing TypeSerializer, old method must have been implemented.
So it could also works even if user don't modify their codes.
If users need to migrate, they could just implement the new method and remove the old method just as `custom_serialization.md` said.
It's implemented just as we talked in https://lists.apache.org/thread/4zgpd4cw232xcjp3cnj2hlokf6zvm3qc

Reply:
> It's implemented just as we talked in https://lists.apache.org/thread/4zgpd4cw232xcjp3cnj2hlokf6zvm3qc

Maybe I missed something there.

Looking at the code. Imagine a situation a user has a snapshot that implements custom logic in `TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(TypeSerializer<T> newSerializer)`.  Our updated code goes through `TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(TypeSerializerSnapshot<T> oldSerializerSnapshot)`, which in this case will return `incompatible`. In result there is no real compatibility and the user is forced to implement the new method.  We may as well simply drop the old method, which will even give you a nicer output as the program won't even compile instead of failing during runtime in (sic!) restore.

Reply:
In milestone 1, the caller still use the old method.
So if a user has a snapshot implements the old method, it will not call the new method.
All behavior will be like before, right ?

Reply:
I don't think this will work. Take this for example:

```
    @Test
    public void testCompatibleWithNonMigratedSerializer() throws IOException {
        TypeSerializer<?>[] testNestedSerializers = {
                new NestedSerializer(TargetCompatibility.COMPATIBLE_AS_IS),
                new NestedSerializer(TargetCompatibility.COMPATIBLE_AS_IS, NestedSerializer.SnapshotType.NOT_MIGRATED),
        };

        TypeSerializerSchemaCompatibility<String> compatibility =
                snapshotCompositeSerializerAndGetSchemaCompatibilityAfterRestore(
                        testNestedSerializers,
                        testNestedSerializers,
                        OuterSchemaCompatibility.COMPATIBLE_AS_IS);

        Assert.assertTrue(compatibility.isCompatibleAsIs());
    }
```
where the `new NestedSerializer(TargetCompatibility.COMPATIBLE_AS_IS, NestedSerializer.SnapshotType.NOT_MIGRATED),` returns a snapshot that still implements `resolveSchemaCompatibility(TypeSerializer<String> newSerializer)` instead of the new method.

This should work, but as it stands it returns `incompatible` forcing user to implement the new method.

Reply:
I think the real reason why previous implementation could not run as we want is that my codes not meet "all callers use the old method" (Composite TypeSerializers will call the new method for their internal TypeSerializers).
I removed migration codes in PojoSerializerSnapshot, CompositeTypeSerializerSnapshot who rely on internal TypeSerializers so that "all callers use the old method".
these classes will be migrated in milestone 2.
Do you think it could work?

Reply:
Could you please add changes after reviews as separate commits? This makes it easier to track what has changed since the last review.

Reply:
What happens if a user has a custom serializer that embeds e.g. a `PojoSerializer` and want to migrate to the new method. A situation like:

```
class CustomCollectionSerializerSnapshot<T> extends TypeSerializerSnapshot<Collection<T>>{
   private final TypeSerializerSnapshot innerSnapshot;
  public ... resolveSchemaCompatibility(TypeSerializerSnapshot oldSnapshot) {
    ...
   return innerSnapshot.resolveSchemaCompatibility(oldSnapshot.innerSnapshot);
 }
}

CustomCollectionSerializer<? extends POJO> serializer
```

This would fail, wouldn't it?

Reply:
Thanks for the detail case. 
It seems it's the most critical problem so I think we could resolve this firstly.

For resolving this, there are some solutions:
1. implement all internal serializers as before pr but not remove the old deprecated method for composite serializers:
  a. Benefit: From users' perspective, previous existing custon serializers and new serializers could work no matter whether users migrate to the new method:
      
      ⅰ. For some previous existing custom serializers that don't migrate, all old code paths will be accessed as before. Users' codes could run without any modifying.

      ⅱ. For some previous existing custom serializers that want to migrate, they could just implement the new method.
          1. If users don't remove the old method, it could also work. Of course, we recomend users to remove it.
          2. If users has used the internal serializers, the new method could just be called.

      ⅲ. For new defined custom serializers:
          1. If users don't implement any methods and restore from the snapshot. INCOMPATIBLE will be returned and the job will fail. Of course, we highly recomend users to implement the new method.
          2. If users implement the new method, it's same as [b].
          
    b. Cost: there are many deprecated method remained.

2. Also implement the new method to call the old method:
  a. Benefit: Then we could just migrate all internal serializers and remove old methods.

    b. Cost: It requires users to implement at least one method else it will stuck in an infinite loop. Although Option 1 also have requirement for users, this is a worse case obviously.

3. Just remove the old method in the interface:
  a. Benefit: it will make code path clear and we could migrate all serializers and make all caller use the new method

    b. Cost: it will beak change so all users have to migrate their serializers to the new method. Else their jobs could not compile.

I think Option 1 looks good to me. WDYT? Or any other suggestions ?

Reply:
Thanks a lot for the detailed explanation.

Reply:
Thanks a lot for the detailed explanation.
I rethinked it and aggreed that option 2 should be a better one currently.